### PR TITLE
[JIT][AD]matmul/dropout

### LINKED ISSUE #17483: [JIT] support range of int
Currently there's no easy way to create [0, 1, 2, ..., x - 1] given x in torchscript, unless we do it in a for loop. 
As discussed offline with @zdevito  and @driazati , we propose adding `prim::range` and let it return a List. 
The behavior doesn't match Python exactly, as range and list in Python are not exactly the same. But this is sufficient for our current need, and I don't see a need to differentiate them atm. 

I have this implemented in a PR I'm working on. Will link to it once that's ready.

### PR DESCRIPTION
**Pull request by ailzhang**
* Add AD formula for [removed: _convolution &]  matmul & dropout
* add prim::range, fixes #17483 
Example: 
```
dim = 3
x = range(dim) 
# x = [0, 1, 2] and is List[int]
```

### CODE REVIEW

**Review by zdevito — APPROVED** [review]
The non-convolution changes look good. In person: let's merge the non-convolution parts of this PR but hold off on changing _convolution. It is not clear dropping really complicated _convolution op into the differentiation will allow any backends to do meaningful optimization.

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -128,6 +128,12 @@ bool isDifferentiable(Node* n) {
   if (differentiable_ops.find(n))
     return true;
 
+  if (n->matches(
+          "aten::dropout(Tensor input, float p, bool train) -> Tensor")) {
+    auto train = n->get<bool>(attr::train).value();
```
Comment (apaszke) [inline torch/csrc/jit/autodiff.cpp]: You never checked that this is a constant. If `train` is a computed value, this will throw. You have to do `n->matches("...", attr::train)` above.

Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: also nit: no need to assign to a local just to return

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -121,6 +121,18 @@ RegisterOperators reg({
             return 0;
           };
         }),
+    Operator(
+        "prim::range(int n) -> int[]",
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: Hmm I'm a bit ambivalent about adding this, since those objects have completely different characteristics in Python.
```
for i in range(2 ** 64):
    ...
```
is fine in Python 3, but we're using Python 2 semantics in TorchScript? That seems like not a bad choice.

File: torch/csrc/jit/symbolic_script.cpp [hunk 3]
```diff
@@ -604,6 +647,18 @@ const std::vector<std::string> functions = {
 
             return output, backward
 
+        def dropout(input,
+                    p: float,
+                    train: bool):
+            mask = torch.empty_like(input)
+            mask.bernoulli_(1 - p)
+            res = mask * input / (1.0 - p)
```
Comment (apaszke) [inline torch/csrc/jit/symbolic_script.cpp]: Why are we replacing forward code with this? It's not necessary, and avoiding it might open up more optimization opportunities and is less bug prone

File: torch/csrc/jit/register_prim_ops.cpp [hunk 4]
```diff
@@ -1113,7 +1125,7 @@ int listSelect<Shared<BoolList>>(Stack& stack) {
   pop(stack, list, idx);
 
   auto element = getBoolItem(list->elements(), idx);
-  push(stack, std::move(element));
+  push(stack, element);
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: Why did we loose the move here? Seems like a regression?

**Comment by apaszke** [comment]
@zdevito FWIW adding convolution might still be nice because we could wrap the whole CNN in a single DifferentiableGraph and save up on constructing autograd graphs (assuming our `DifferentiableGraphFunction` is fast enough). But it's a lot of work.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (ailzhang) [reply torch/csrc/jit/register_prim_ops.cpp]: Yea I wasn't sure about this but it seems very useful atm as it saves me from writing for unnecessary loops in torchscript. 
An alternative of this I thought about was list comprehension which in general I think it's also super useful, I went for prim::range because it's  quick to implement I think. Possibly we could not document this feature and remove it when list comprehension is ready?

File: torch/csrc/jit/symbolic_script.cpp (hunk 3 above)
Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: When I wrote this I had 2 options : 
- changing https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Dropout.cpp#L54 to return the mask as well 
- write a simple dropout implementation in torchscript. 
I went for the latter solution mainly because it's straightforward to do with a few lines. 

I think a design question here is : do we want to keep AD formula as close as possible to its ATen counterpart (for easier possible future deduplication and consistency), or do we want to keep AD only math and simple? In some cases our aten functions considers backend/special sizes etc, I'm not certain if we would like to include those in AD formula or not. Let me know your thoughts on this, would love to have some guidelines on it. Thanks!

File: torch/csrc/jit/register_prim_ops.cpp (hunk 4 above)
Comment (ailzhang) [reply torch/csrc/jit/register_prim_ops.cpp]: Hmmm iirc this is done by clang-tidy hook. It was suggesting moving a bool doesn't make a perf difference and I thought that makes sense. I could bring it back if this might cause regression.

Comment (apaszke) [reply torch/csrc/jit/register_prim_ops.cpp]: Ohhh I haven't noticed it's a bool. Yeah, in that case it shouldn't be there

File: torch/csrc/jit/symbolic_script.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/symbolic_script.cpp]: Fair enough, I forgot that you need the mask! Let's keep it like that.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/register_prim_ops.cpp]: I'm fine with adding a special primitive, just don't call it `range`. If we use Python names we should match Python semantics.

Comment (ailzhang) [reply torch/csrc/jit/register_prim_ops.cpp]: Yea that makes sense, I will rename it in a following PR. :) Thanks!
